File: java/src/main/java/io/binac/leetcode/ReverseInteger.java
package io.binac.leetcode;

/**
 * <p>Given a signed 32-bit integer <code>x</code>, return <code>x</code><em> with its digits reversed</em>. If reversing <code>x</code> causes the value to go outside the signed 32-bit integer range <code>[-2<sup>31</sup>, 2<sup>31</sup> - 1]</code>, then return <code>0</code>.</p>
 *
 * <p><strong>Assume the environment does not allow you to store 64-bit integers (signed or unsigned).</strong></p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 * <pre><strong>Input:</strong> x = 123
 * <strong>Output:</strong> 321
 * </pre><p><strong>Example 2:</strong></p>
 * <pre><strong>Input:</strong> x = -123
 * <strong>Output:</strong> -321
 * </pre><p><strong>Example 3:</strong></p>
 * <pre><strong>Input:</strong> x = 120
 * <strong>Output:</strong> 21
 * </pre><p><strong>Example 4:</strong></p>
 * <pre><strong>Input:</strong> x = 0
 * <strong>Output:</strong> 0
 * </pre>
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li><code>-2<sup>31</sup> &lt;= x &lt;= 2<sup>31</sup> - 1</code></li>
 * </ul>
 */
public class ReverseInteger {
    public static class Solution1 {
        public int reverse(int x) {
            int result = 0;
            for (; x != 0; x /= 10) {
                if (result > 214748364 || result < -214748364)
                    return 0;
                result = result * 10 + x % 10;
            }
            return result;
        }
    }
}

File: python3/leetcodepy/reverse_integer.py
"""
Given a signed 32-bit integer x, return x with its digits reversed. If reversing x causes the value to go outside the signed 32-bit integer range [-2^31, 2^31 - 1], then return 0.

Assume the environment does not allow you to store 64-bit integers (signed or unsigned).



Example 1:

Input: x = 123
Output: 321

Example 2:

Input: x = -123
Output: -321

Example 3:

Input: x = 120
Output: 21

Example 4:

Input: x = 0
Output: 0


Constraints:

-2^31 <= x <= 2^31 - 1
"""


class Solution1:
    def reverse(self, x: int) -> int:
        result = 0
        mod = 10 if x >= 0 else -10  # correction for python
        while x != 0:
            if result > 214748364 or result < -214748364:
                return 0
            result = result * 10 + x % mod
            x = int(x / 10)
        return result
